File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    visited = [False] * len(dungeons)
    def dfs(curHp, cnt):
        if cnt < len(dungeons):
            max_cnt = cnt
            for idx in range(len(dungeons)):
                if curHp >= dungeons[idx][0] and not visited[idx]:
                    visited[idx] = True
                    max_cnt = max(max_cnt, dfs(curHp - dungeons[idx][1], cnt + 1))
                    visited[idx] = False
            return max_cnt
        return cnt
    
    return dfs(k,0)

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    var answer = 0;
    let visited = new Array(dungeons.length).fill(false);
    
    function dfs(curHP, cnt) {
        
        for (let i = 0; i < dungeons.length; i ++) {
            if (curHP >= dungeons[i][0] && !visited[i]) {
                visited[i] = true
                dfs(curHP - dungeons[i][1], cnt + 1)
                visited[i] = false
            }
        }
        
        answer = Math.max(answer, cnt)
    }
    
    dfs(k, 0);
    return answer;
}
